### DISCUSSION TITLE
I want to find opcode using ghidra using headless.

### DISCUSSION THREAD

**Discussion by anzosasuke** [NationalSecurityAgency/ghidra#3021] [General]
II want to find opcode using ghidra headless. Is there any way to do that??

**Comment by astrelsky** [comment]
>  II want to find opcode using ghidra headless. Is there any way to do that??

Do you mean you want to find all instances of a specific instruction?

If so you could brute force it with something like:

```python

def get_inst(mnemonic):
    i = getFirstInstruction()
    while i:
        monitor.checkCanceled()
        if i.getMnemonicString() == mnemonic:
            yield i
        i = getInstructionAfter(i)
```

You could improve this by further filtering it by the pcodeop your instruction produces before checking if the mnemonic matches. Iirc there is an instruction search service but in headless mode there is no tool which means no way to get the service.

**Reply by anzosasuke** [reply]
Thank you for your response. I should have worded it properly, I wanted an opcode byte. Like there is a instruction which has opcode + operand. That mnemonic's actual binary/hex representation in the machine. I hope I made you understand.

**Reply by astrelsky** [reply]
> Thank you for your response. I should have worded it properly, I wanted an opcode byte. Like there is a instruction which has opcode + operand. That mnemonic's actual binary/hex representation in the machine. I hope I made you understand.

I see. Here are the javadocs for [Instruction](https://ghidra.re/ghidra_docs/api/ghidra/program/model/listing/Instruction.html). It implements [CodeUnit](https://ghidra.re/ghidra_docs/api/ghidra/program/model/listing/CodeUnit.html) so you can use [CodeUnit.getBytes()](https://ghidra.re/ghidra_docs/api/ghidra/program/model/listing/CodeUnit.html#getBytes()). You can iterate and search for the instructions using the same method from before. Additionally if you already know what the bytes are there are  `findBytes` methods in the `FlatProgramAPI`. Iirc there is also `DittedBitSequence` or something like that. Its what is used to find functions from the processor specific function pattern files.

**Reply by anzosasuke** [reply]
Thank you so much. I got the bytes, codeUnit.getBytes worked well.
